spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/KafkaStreamsFunctionProcessor.java

```diff
@@ -107,9 +108,38 @@ public KafkaStreamsFunctionProcessor(BindingServiceProperties bindingServiceProp
 	}
 
 	private Map<String, ResolvableType> buildTypeMap(ResolvableType resolvableType,
-													KafkaStreamsBindableProxyFactory kafkaStreamsBindableProxyFactory) {
+													KafkaStreamsBindableProxyFactory kafkaStreamsBindableProxyFactory,
+													Method method, String functionName) {
 		Map<String, ResolvableType> resolvableTypeMap = new LinkedHashMap<>();
-		if (resolvableType != null && resolvableType.getRawClass() != null) {
+		if (method != null) { // Component functional bean.
+			final ResolvableType firstMethodParameter = ResolvableType.forMethodParameter(method, 0);
+			ResolvableType currentOutputGeneric = ResolvableType.forMethodReturnType(method);
+
+			final Set<String> inputs = new LinkedHashSet<>(kafkaStreamsBindableProxyFactory.getInputs());
+
+			final Iterator<String> iterator = inputs.iterator();
+
+			popuateResolvableTypeMap(firstMethodParameter, resolvableTypeMap, iterator, method, functionName);
+
+			final Class<?> outputRawclass = currentOutputGeneric.getRawClass();
+			if (outputRawclass != null && !outputRawclass.equals(Void.TYPE)) {
+				ResolvableType iterableResType = currentOutputGeneric;
+				int i = 1;
+				while (i < inputs.size() && iterator.hasNext()) {
```

Comment:
Soby, personally I would put this logic in some private method with distinctive name and may be some comments to improve readability